dimod/discrete/discrete_quadratic_model.py

```diff
@@ -810,3 +808,315 @@ def to_numpy_vectors(self, return_offset: bool = False):
+                the labels that will identify the cases of the variable.
+
+            label (hashable, optional):
+                A label for the variable. Can be any hashable except `None`.
+                Defaults to the length of the discrete quadratic model, if that
+                label is available. Otherwise defaults to the lowest available
+                positive integer label.
+
+            shared_labels (bool, optional, default=False):
+                If True and `cases` is an iterable, shared case labels are
+                created.  If False and `cases` is an iterable, unique case
+                labels are created.  If `cases` is not an iterable, ignored.
+
+        Returns:
+            The label of the new variable.
+
+        Raises:
+            ValueError: If `label` already exists as a variable label, or if
+                any of the case labels is not unique.
+
+            TypeError: If `label` is not hashable, or if any of the case labels
+                is not hashable.
+
+        """
+        if label in self._unique_label_case:
+            raise ValueError(f'variable label {label} is not unique')
+
+        if isinstance(cases, int):
+            return super().add_variable(cases, label=label)
+
+        else:
+            if len(set(cases)) != len(cases):
+                raise ValueError('case labels are not unique')
+
+            if shared_labels:
+                var = super().add_variable(len(cases), label=label)
+
+                for k, case in enumerate(cases):
+                    self._shared_label_case[var][case] = k
+                    self._shared_case_label[var][k] = case
+
+            else:
+                for case in cases:
+                    if (case in self.variables) or (case in self._unique_label_case):
+                        raise ValueError(f'case label {case} is not unique')
+
+                var = super().add_variable(len(cases), label=label)
+
+                for k, case in enumerate(cases):
+                    self._unique_label_case[case] = (var, k)
+                    self._unique_case_label[(var, k)] = case
+
+            return var
+
+    def _lookup_shared_case(self, v, case):
+        """Translate shared case label `case` of variable `v` to integer.
+
+        Raises:
+            ValueError: If `case` of `v` is unknown.
+
+        """
+        map_ = self._shared_label_case.get(v)
+        if map_:
+            if case not in map_:
+                raise ValueError(f'unknown case {case} of variable {v}')
+            return map_[case]
+        return case
+
+    def get_linear(self, v):
+        """The linear bias(es) associated with variable `v`.
+
+        Args:
+            v: A variable in the discrete quadratic model, or a unique case
+                label.
+
+        Returns:
+            The linear bias(es).  If `v` is a variable, returns a
+                :meth:`~DiscreteQuadraticModel.num_cases(v)` by 1 numpy array.
```

Comment:
"numpy" --> Numpy
Would be easier to parse as "... returns a Numpy array of size :meth:`~DiscreteQuadraticModel.num_cases(v)` by 1"
(Also elsewhere below)

Reply:
Just to note I see we use "numpy" in three pre-existing doc strings: `DQM.to_numpy_vectors()`, `DQM.from_numpy_vectors()`, and `DQM.get_quadratic()`, so this will introduce a minor inconsistency, unless we change them as well.

Reply:
We're inconsistent about this all over Ocean. `NumPy` is actually how they refer to themselves so I think we should use that when we remember.

Reply:
Please do. And it's actually NumPy (sorry, my typo)

Reply:
I don't think that change should be made as part of this PR.

Reply:
Fixed.